modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -315,6 +353,142 @@ def find_esc3_vuln_cert_templates
     query_ldap_server_certificates(esc3_template_2_raw_filter, 'ESC3_TEMPLATE_2')
   end
 
+  def find_esc4_vuln_cert_templates
+    # Determine who we are authenticating with. Retrieve the username and user SID
+    whoami_response = ''
+    begin
+      whoami_response = @ldap.ldapwhoami
+    rescue Net::LDAP::Error => e
+      print_warning("The module failed to run the ldapwhoami command, ESC4 detection can't continue. Error was: #{e.class}: #{e.message}.")
+      return
+    end
+
+    if whoami_response.empty?
+      print_error("Unable to retrieve the username using ldapwhoami, ESC4 detection can't continue")
+      return
+    end
+
+    sam_account_name = whoami_response.split('\\')[1]
+    user_raw_filter = "(sAMAccountName=#{sam_account_name})"
+    attributes = ['objectSID', 'DN']
+    our_user = query_ldap_server(user_raw_filter, attributes)&.first
+    if our_user.nil?
+      print_warning("Unable to determine the User SID for #{sam_account_name}, ESC4 detection can't continue")
+      return
+    else
+      user_sid = Rex::Proto::MsDtyp::MsDtypSid.read(our_user[:objectsid].first).value
+    end
+
+    # Determine if we are a User or a Computer
+    fiter_user_or_computer = "(objectSid=#{user_sid})"
+    attributes = ['objectClass']
+    response_computer_or_user = query_ldap_server(fiter_user_or_computer, attributes)&.first
+
+    if response_computer_or_user.nil?
+      print_warning("Unable to determine the if SID belongs to a user account or a computer account, ESC4 detection can't continue")
+      return
+    end
+
+    if response_computer_or_user[:objectclass].include?('computer')
```

Comment:
We should be able to cut down one LDAP query here by adding the `objectClass` to the attributes we want when looking up the account. I guess `our_user` now could be a bit misleading of a variable name, maybe `our_account` would be better even though *technically* computer accounts still have the `user` objectclass as you noticed.

```suggestion
    sam_account_name = whoami_response.split('\\')[1]
    user_raw_filter = "(sAMAccountName=#{sam_account_name})"
    attributes = ['DN', 'objectSID', 'objectClass']
    our_user = query_ldap_server(user_raw_filter, attributes)&.first
    if our_user.nil?
      print_warning("Unable to determine the User SID for #{sam_account_name}, ESC4 detection can't continue")
      return
    end
    
    user_sid = Rex::Proto::MsDtyp::MsDtypSid.read(our_user[:objectsid].first).value

    if our_user[:objectclass].include?('computer')
```